storm/src/main/storm/mesos/util/ZKClient.java

```diff
@@ -0,0 +1,117 @@
+package storm.mesos.util;
+
+import org.apache.curator.RetryPolicy;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.apache.curator.framework.api.BackgroundCallback;
+import org.apache.curator.framework.api.CuratorEvent;
+import org.apache.curator.framework.api.CuratorListener;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * ZKClient allows you to interact with ZooKeeper. Primarily used for tracking logviewer state on hosts thus far but
+ * can be used to track any metadata needed in the future.
+ */
+
+public class ZKClient {
+  CuratorFramework _client;
+  public static final Logger LOG = LoggerFactory.getLogger(ZKClient.class);
+  private static final int BASE_SLEEP_TIME_MS = 1000;
+  private static final int MAX_RETRIES = 3;
+
+  public ZKClient() {
+    ExponentialBackoffRetry retryPolicy = new ExponentialBackoffRetry(BASE_SLEEP_TIME_MS, MAX_RETRIES);
+    _client = CuratorFrameworkFactory.newClient("localhost:2181", retryPolicy);
+    _client.start();
+  }
+
+  public ZKClient(String connectionString) {
+    LOG.info("Attempting to connect to following ZooKeeper servers: {}", connectionString);
+    ExponentialBackoffRetry retryPolicy = new ExponentialBackoffRetry(BASE_SLEEP_TIME_MS, MAX_RETRIES);
+    _client = CuratorFrameworkFactory.newClient(connectionString, retryPolicy);
+    _client.start();
+  }
+
+  public ZKClient(String connectionString, int connectionTimeout, int sessionTimeout) {
+    LOG.info("Attempting to connect to following ZooKeeper servers: {}", connectionString);
+    ExponentialBackoffRetry retryPolicy = new ExponentialBackoffRetry(BASE_SLEEP_TIME_MS, MAX_RETRIES);
+    _client = CuratorFrameworkFactory.builder()
+                                     .connectString(connectionString)
+                                     .retryPolicy(retryPolicy)
+                                     .connectionTimeoutMs(connectionTimeout)
+                                     .sessionTimeoutMs(sessionTimeout)
+                                     .build();
+    _client.start();
+  }
+
+  public boolean createNode(String path, String data) {
+    try {
+      _client.create().creatingParentsIfNeeded().forPath(path, data.getBytes());
+      return true;
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      return false;
+    }
+  }
+
+  public boolean createNode(String path) {
+    try {
+      // default payload of byte[0]
+      _client.create().creatingParentsIfNeeded().forPath(path);
+      return true;
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      return false;
+    }
+  }
+
+  public void deleteNode(String path) {
+    try {
+      // the delete is guaranteed even if initial failure, curator will attempt to delete in background until successful
+      _client.delete().guaranteed().forPath(path);
```

Comment:
@erikdw If there is a logviewer task running that ZK doesn't know about that will be eventually reconciled. But currently, if ZK thinks there is a logviewer running but it is actually dead (aka if the Nimbus dies then the logviewer dies) then we will not reconcile that state because we don't have "heartbeats"

SUGGESTION: If we track the TaskIDs of logviewers somewhere (ZK, List<TaskID>, etc.) then we can reconcile that state (through task reconciliation) by passing it the logviewer TaskIDs from which we would receive TASK_LOST if they were actually dead.

Reply:
@changreytang : thanks for implementing the SUGGESTION, per our offline discussions!